clang/lib/Frontend/CompileJobCacheResult.cpp

```diff
@@ -34,6 +34,39 @@ Error CompileJobCacheResult::forEachOutput(
   return Error::success();
 }
 
+Error CompileJobCacheResult::forEachLoadedOutput(
+    llvm::function_ref<Error(Output, std::optional<ObjectProxy>)> Callback) {
+  // Kick-off materialization for all outputs concurrently.
+  SmallVector<std::future<Expected<std::optional<ObjectProxy>>>, 4>
+      FutureOutputs;
+  size_t Count = getNumOutputs();
+  for (size_t I = 0; I < Count; ++I) {
+    ObjectRef Ref = getOutputObject(I);
+    FutureOutputs.push_back(getCAS().getProxyAsync(Ref));
+  }
+
+  // Make sure all the outputs have materialized.
+  std::optional<Error> OccurredError;
+  SmallVector<std::optional<ObjectProxy>, 4> Outputs;
+  for (auto &FutureOutput : FutureOutputs) {
+    auto Obj = FutureOutput.get();
+    if (!Obj) {
+      OccurredError = Obj.takeError();
```

Comment:
This looks wrong if there are multiple errors; only the last one will be considered "checked".  I think you want something like

```c++
if (!OccurredError)
  OccurredError = Obj.takeError();
else
  OccurredError = llvm::joinErrors(*OccurredError, Obj.takeError());
```

Reply:
> This looks wrong if there are multiple errors; only the last one will be considered "checked"

I'm checking the error right before moving it to `OccurredError`, why is it still considered "unchecked"?
Also the "unchecked" assertion fires even when there is no error, so this should have been failing if that was the case.

Reply:
No, doing `if` is only good enough to "check" if the error is "success":
```
  explicit operator bool() {
    setChecked(getPtr() == nullptr);
    return getPtr() != nullptr;
  }
```

If there is an actual error you need to handle it.

Reply:
I see, good catch!

Reply:
Fixed here: https://github.com/apple/llvm-project/pull/6458/commits/420e373ffb4138fce3746977dbf4bde71aa7fb99